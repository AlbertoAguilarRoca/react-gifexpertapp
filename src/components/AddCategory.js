import React, { useState } from 'react';
import PropTypes from 'prop-types';

//recibe por parametro la funcion del hook useState
export const AddCategory = ( {setCategories} ) => {
    //hook useState con la informacion del input 
    const [inputValue, setinputValue] = useState('');

    const handleInputValue = (e) => {
        setinputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        
        //metemos dentro del useState la nueva categoria
        if(inputValue.trim().length > 2) {
            //Metemos el nuevo valor junto con las categorias
            //que ya habia antes
            setCategories(category => [inputValue, ...category ]);
            setinputValue('');
        }
    }

    return (
        <form onSubmit={handleSubmit}>
            <div className='buscador'>
                <i className='bi bi-plus-circle'></i>
                <input 
                    type='text' 
                    value={ inputValue }
                    onChange = { handleInputValue }
                />
            </div>
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
