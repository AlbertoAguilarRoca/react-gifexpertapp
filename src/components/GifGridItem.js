import React from 'react'
import PropTypes from 'prop-types';

export const GifGridItem = ({title, url}) => {

  return (
    <div className='card animate__animated animate__fadeInDown'>
      <div className='card__content'>
        <img src={url} alt={title} />
        <h3>{title}</h3>
      </div>
    </div>
  )
}

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
}
