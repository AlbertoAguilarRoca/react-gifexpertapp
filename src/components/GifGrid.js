import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem'

//se recibe la categoria que se va a buscar
export const GifGrid = ({category}) => {

    const {data:images, loading} = useFetchGifs(category);

    //imprimira las imagenes a traves de un nuevo componente
    //en vez de enviar un parametro, se envia ...img para poder
    //desestructurar dicho componente
    //Cuando el componente se ejecuta varias veces, hay que poner una key unica
    return (
        <>
            <h2 className='animate__animated animate__fadeInDown'>{category}</h2>

            {loading && <p className='animate__animated animate__fadeInDown'>
                Cargando...</p>}

            <div className='card__row'>
                {
                    images.map( img => (
                        <GifGridItem 
                            key = {img.id}
                            {...img} 
                        />
                    ))
                }
            </div>
        </>
    )
}
