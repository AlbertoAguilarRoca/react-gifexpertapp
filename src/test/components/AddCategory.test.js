import React from 'react';
import {shallow} from 'enzyme';
import { AddCategory } from '../../components/AddCategory';


describe('Pruebas en <AddCategory />>', () => { 

    //jest.fn() simula una funcion cualquiera
    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories} />);

    beforeEach(() => {
        //Todas las funciones que se me han mockeado se limpian
        //antes de cada test
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories} />);
    });

    test('debe de mostrarse correctamente', () => { 
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de cambiar la caja de texto', () => {
        const input = wrapper.find('input');
        const value = 'Goku';
        //Ese target me permite simular el parametro e
        //que se le pasa a la funcion al lanzar el evento
        input.simulate('change', {target: {value}});
    });

    test('No debe de postear la informacion al hacer submit', () => {
        wrapper.find('form').simulate('submit', {
            preventDefault(){} //mandamos un metodo a traves del evento
        });
        //Al no tener contenido el inputValue, no se debe de lanzar la
        //funcion setCategory. Eso es lo que vamos a testar
        expect(setCategories).not.toHaveBeenCalled();
    });

    test('Debe de llamar el setCategories y limpiar la caja de texto', () => {
        const input = wrapper.find('input');
        const value = 'Goku';
        //simulo el change
        input.simulate('change', {target: {value}});
        //simulo el submit
        wrapper.find('form').simulate('submit', {
            preventDefault(){} //mandamos un metodo a traves del evento
        });
        expect(setCategories).toHaveBeenCalled();
        expect(input.prop('value')).toBe('');
    });

});