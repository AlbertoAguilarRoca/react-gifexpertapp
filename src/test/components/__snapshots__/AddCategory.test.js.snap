// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory />> debe de mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="buscador"
  >
    <i
      className="bi bi-plus-circle"
    />
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
</form>
`;
